core/src/main/java/org/apache/brooklyn/core/objs/AbstractBrooklynObject.java

```diff
@@ -51,7 +55,7 @@
     @SetFromFlag("id")
     private String id = Identifiers.makeRandomLowercaseId(10);
 
-    private String catalogItemId;
+    private Deque<String> catalogItemIdStack = new ArrayDeque<>();
```

Comment:
`BrooklynObject`s get created by calling their constructors so removing `catalogItemId` and initializing `catalogItemIdStack`  should be safe.